rust/test-lsp-wake/src/main.rs

```diff
@@ -0,0 +1,287 @@
+    let length = str.len();
+
+    format!("Content-Length: {}\r\n\r\n{}", length, str)
+}
+
+fn make_request<R>(id: tower_lsp::jsonrpc::Id, params: R::Params) -> tower_lsp::jsonrpc::Request
+where
+    R: tower_lsp::lsp_types::request::Request,
+    R::Params: serde::Serialize,
+    R::Result: serde::de::DeserializeOwned,
+{
+    tower_lsp::jsonrpc::Request::build(R::METHOD)
+        .id(id)
+        .params(serde_json::to_value(params).expect(""))
+        .finish()
+}
+
+async fn write_request<R>(
+    stdin: &mut ChildStdin,
+    id: tower_lsp::jsonrpc::Id,
+    params: R::Params,
+) -> Result<usize, std::io::Error>
+where
+    R: tower_lsp::lsp_types::request::Request,
+    R::Params: serde::Serialize,
+    R::Result: serde::de::DeserializeOwned,
+{
+    let req = make_request::<R>(id, params);
+    stdin.write(jrpc_serialize(req).as_bytes()).await
+}
+
+fn make_notification<R>(params: R::Params) -> tower_lsp::jsonrpc::Request
+where
+    R: tower_lsp::lsp_types::notification::Notification,
+    R::Params: serde::Serialize,
+{
+    tower_lsp::jsonrpc::Request::build(R::METHOD)
+        .params(serde_json::to_value(params).expect(""))
+        .finish()
+}
+
+async fn write_notification<R>(
+    stdin: &mut ChildStdin,
+    params: R::Params,
+) -> Result<usize, std::io::Error>
+where
+    R: tower_lsp::lsp_types::notification::Notification,
+    R::Params: serde::Serialize,
+{
+    let req = make_notification::<R>(params);
+    stdin.write(jrpc_serialize(req).as_bytes()).await
+}
+
+async fn read_notification<R>(
+    reader: &mut Lines<BufReader<ChildStdout>>,
+) -> std::io::Result<R::Params>
+where
+    R: tower_lsp::lsp_types::notification::Notification,
+    R::Params: serde::Serialize,
+{
+    let Some(_length) = reader.next_line().await? else {
+        todo!();
+    };
+    let Some(_end_header) = reader.next_line().await? else {
+        todo!();
+    };
+    let Some(json) = reader.next_line().await? else {
+        todo!();
+    };
+
+    let res: tower_lsp::jsonrpc::Request = serde_json::from_str(&json).unwrap();
+    let value = res.params().unwrap();
+    let inner: R::Params = serde_json::from_value(value.clone()).unwrap();
+    return Ok(inner);
+}
+
+async fn read_response<R>(reader: &mut Lines<BufReader<ChildStdout>>) -> std::io::Result<R::Result>
+where
+    R: tower_lsp::lsp_types::request::Request,
+    R::Params: serde::Serialize,
+    R::Result: serde::de::DeserializeOwned,
+{
+    let _length = reader.next_line().await?;
+    let _end_header = reader.next_line().await?;
+    let Some(json) = reader.next_line().await? else {
+        todo!();
+    };
+
+    let res: tower_lsp::jsonrpc::Response = serde_json::from_str(&json).unwrap();
+    let value = res.result().unwrap();
+    let inner: R::Result = serde_json::from_value(value.clone()).unwrap();
+    return Ok(inner);
+}
+
+async fn exchange_request_response<R>(
+    stdin: &mut ChildStdin,
+    reader: &mut Lines<BufReader<ChildStdout>>,
```

Comment:
I think you should put these two things inside of a struct and make these impl methods instead.